clients/src/main/java/org/apache/kafka/common/security/ssl/CommonNameLoggingTrustManagerFactoryWrapper.java

```diff
@@ -0,0 +1,423 @@
+     * or rethrow the original exception if it is not.
+     */
+    static class CommonNameLoggingTrustManager implements X509TrustManager {
+
+        final private X509TrustManager origTm;
+        final int nrOfRememberedBadCerts;
+        final private LinkedHashMap<String, String> previouslyRejectedClientCertChains;
+
+        public CommonNameLoggingTrustManager(X509TrustManager originalTrustManager, int nrOfRememberedBadCerts) {
+            this.origTm = originalTrustManager;
+            this.nrOfRememberedBadCerts = nrOfRememberedBadCerts;
+            // Restrict maximal size of the LinkedHashMap to avoid security attacks causing OOM
+            this.previouslyRejectedClientCertChains = new LinkedHashMap<String, String>() {
+                @Override
+                protected boolean removeEldestEntry(final Map.Entry<String, String> eldest) {
+                    return size() > nrOfRememberedBadCerts;
+                }
+            };
+        }
+
+        public X509TrustManager getOriginalTrustManager() {
+            return this.origTm;
+        }
+
+        @Override
+        public void checkClientTrusted(X509Certificate[] chain, String authType)
+                throws CertificateException {
+            CertificateException origException = null;
+            String chainDigest = calcDigestForCertificateChain(chain);
+            if (chainDigest != null) {
+                String errorMessage = this.previouslyRejectedClientCertChains.get(chainDigest);
+                if (errorMessage != null) {
+                    // Reinsert the digest, to remember that this is the most recent digest we've seen
+                    addRejectedClientCertChains(chainDigest, errorMessage, true);
+                    // Then throw with the original error Message
+                    throw new CertificateException(errorMessage);
+                }
+            }
+            try {
+                this.origTm.checkClientTrusted(chain, authType);
+                // If the last line did not throw, the chain is valid (including that none of the certificates is expired)
+            } catch (CertificateException e) {
+                origException = e;
+                try {
+                    X509Certificate[] wrappedChain = sortChainAnWrapEndCertificate(chain);
+                    this.origTm.checkClientTrusted(wrappedChain, authType);
+                    // No exception occurred this time. The certificate is either not yet valid or already expired
+                    Date now = new Date();
+                    // Check if the certificate was valid in the past
+                    if (wrappedChain[0].getNotBefore().before(now)) {
+                        String commonName = wrappedChain[0].getSubjectX500Principal().toString();
+                        String notValidAfter = wrappedChain[0].getNotAfter().toString();
+                        log.info("Certificate with common name \"" + commonName + "\" expired on " + notValidAfter);
```

Comment:
Actually, I have not considered this, for three reasons. First, we made the observation that broker-side monitoring is in general much more frequently implemented than client-side monitoring, due to the higher complexity of the latter. Second, if a devops team has proper client-side monitoring in place, they will figure out quickly that their client fails and what the reason is. The proposed change is meant to support teams responsible for the Kafka platform to identify misconfigured clients and contact the responsible developers. Third, it might be seen as a security flaw to provide too much details to clients why exactly their connection has been rejected as there might be an attack on the broker ongoing.